File: 03월 11일 - 스택, 큐, 덱/과제/2108.cpp
//
// Created by JIYOUNG.P on 2022-03-10.
//

#include <iostream>
#include <vector>
#include <algorithm>    // sort() 사용
#include <numeric>      // accumulate() 사용
#include <cmath>        // floor() || round() 사용

using namespace std;

typedef pair<int, int> ci;  // pair<int, int>에 ci라는 별칭 부여

// 비교함수
bool cmp(const ci &a, const ci &b) {
    if (a.first == b.first) {
        return a.second < b.second;
    }
    return a.first > b.first;
}

int findMode(int n, vector<int> &arr) {
    // 정렬된 벡터에서 최빈값을 찾아 반환하고, 만약 최빈값이 여러개라면 두번째로 작은 값을 반환하는 함수
    vector<ci> count;   // 정수의 개수와 해당 정수 값을 pair로 묶어서 저장하는 벡터
    // first: 개수, second: 값

    int current_idx = 0;    // 현재의 인덱스를 저장하는 변수
    count.push_back({1, arr[0]});   // 첫번째 값을 입력하여 for문에서 index 에러 방지

    for (int i = 1; i < n; i++) {
        // 만약 직전 값과 같은 값이라면
        if (arr[i] == arr[i - 1]) {
            // 개수 +1
            count[current_idx].first++;
        } else {
            // 그렇지 않다면, 새로운 값을 count 벡터에 삽입
            count.push_back({1, arr[i]});
            current_idx++;
        }
    }

    // 만약 값이 한 종류라면, 바로 리턴 -> 아래에서 인덱스 에러 방지
    if (current_idx == 0) {
        return arr[0];
    }

    // 정렬
    // 1. 개수(first)에 대해 내림차순
    // 2. 값(second)에 대해 오름차순
    sort(count.begin(), count.end(), cmp);

    // 최빈 값이 여러개인지 확인
    if (count[0].first == count[1].first) {
        return count[1].second;
    }
    return count[0].second;
}

int main() {
    int n;
    cin >> n;

    // n 최대 500,000 * 입력값 최대 4,000 = 2,000,000,000 <- int 범위 안
    vector<int> arr(n);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    sort(arr.begin(), arr.end());
    // 산술평균 구하기, 나눠주기 위해서 float형으로 받아야 합니다.
    float sum = accumulate(arr.begin(), arr.end(), 0);
    cout << floor(sum / n + 0.5) << '\n';

    // 중앙값 구하기
    cout << arr[n / 2] << '\n';

    // 최빈값 구하기
    cout << findMode(n, arr) << '\n';

    // 범위 구하기
    cout << arr[n - 1] - arr[0] << '\n';
}

File: 03월 11일 - 스택, 큐, 덱/과제/2108.py
import sys
from math import floor  # round 함수를 대체할 floor(내림)함수
input = sys.stdin.readline

def find_mode(n, arr):
    # 정렬된 벡터에서 최빈값을 찾아 반환하고, 만약 최빈값이 여러개라면 두번째로 작은 값을 반환하는 함수
    count = [] # [정수의 값, 정수의 개수]으로 구성된 리스트

    current_idx = 0 # 현재 인덱스를 저장하는 변수
    count.append([arr[0], 1])   # 첫번째 값을 미리 입력하여 for문에서 index 에러 방지
    # 주의! 이때 튜플을 사용하면, 값을 업데이트 할 수 없음

    for i in range(1, n):
        # 만약 직전 값과 같은 값이라면
        if arr[i] == arr[i-1]:
            # 개수 + 1
            count[current_idx][1] += 1
        else:
            # 그렇지 않다면, 새로운 값을 count리스트에 추가
            count.append([arr[i], 1])
            current_idx += 1

    # 만약 값이 한 종류라면, 바로 리턴 -> 아래에서 인덱스 에러 방지
    if current_idx == 0:
        return arr[0]

    # 정렬
    # 1. 개수에 대해 내림차순
    # 2. 값에 대해 오름차순
    count.sort(key=lambda x:(-x[1], x[0]))

    # 최빈값이 여러개인지 확인
    if count[0][1] == count[1][1]:
        return count[1][0]
    
    return count[0][0]


def print_values(n, arr):
    # 산술평균
    # python의 round 함수는 사사오입 원칙을 따릅니다.
    # 즉, 반올림 할 숫자가 5이면 앞자리를 짝수로 만들도록 올림/내림합니다.
    # 따라서, 우리가 원하는 출력을 하기 위해서는 round 함수를 쓰면 안됩니다.
    mean = sum(arr) / n
    print(floor(mean + 0.5))    # 0.5를 더해 내림해주면, 우리가 생각한대로 4까지는 내림, 5부터는 올림 연산을 수행하게 됩니다.

    # 중앙값
    print(arr[n//2])

    # 최빈값
    print(find_mode(n, arr))

    # 범위
    print(arr[-1] - arr[0])


n = int(input())
arr = [int(input()) for _ in range(n)]
arr.sort()

print_values(n, arr)